snuba/cli/query_fetcher.py

```diff
@@ -0,0 +1,210 @@
+import csv
+from collections import namedtuple
+from datetime import datetime, timedelta
+from typing import Optional, Sequence, Tuple
+
+import click
+import structlog
+
+from snuba.clickhouse.native import ClickhousePool
+from snuba.clusters.cluster import ClickhouseClientSettings
+from snuba.environment import setup_logging, setup_sentry
+from snuba.utils.gcs import GCSUploader
+
+logger = structlog.get_logger().bind(module=__name__)
+
+QueryToRerun = namedtuple("QueryToRerun", ["query_str", "query_id"])
+
+
+class QueriesFileSaver:
+    def __init__(self, gcs_bucket: Optional[str] = None) -> None:
+        self.uploader = None
+        if gcs_bucket:
+            try:
+                self.uploader = GCSUploader(gcs_bucket)
+            except Exception as e:
+                logger.info(f"Couldn't set up gcs for bucket {gcs_bucket}: {str(e)}")
+
+    def _format_filename(self, table: str, date: datetime) -> str:
+        # Example: queries_2024_01_16_1_errors_local - first hour
+        #          queries_2024_01_16_2_errors_local- second hour
+        day = datetime.strftime(date, "%Y_%m_%d")
+        hour = date.hour
+        return f"queries_{day}_{hour}_{table}"
+
+    def _format_blob_name(self, table: str, date: datetime) -> str:
+        # Example: queries/2024_01_16/1/errors_local - first hour
+        #          queries/2024_01_16/2/errors_local- second hour
+        day = datetime.strftime(date, "%Y_%m_%d")
+        hour = date.hour
+        return f"queries/{day}/{hour}/{table}"
+
+    def _full_path(self, filename: str) -> str:
+        return f"/tmp/{filename}.csv"
+
+    def _save_to_csv(self, filename: str, results: Sequence[QueryToRerun]) -> None:
+        with open(self._full_path(filename), mode="w") as file:
+            writer = csv.writer(file)
+            for row in results:
+                writer.writerow(row)
+
+        logger.info(f"File {self._full_path(filename)} saved")
+
+    def _save_to_gcs(self, filename: str, blob_name: str) -> None:
+        if self.uploader:
+            self.uploader.upload_file(self._full_path(filename), blob_name)
+
+    def save(self, table: str, date: datetime, results: Sequence[QueryToRerun]) -> None:
+        """
+        First save the results to local csv file,
+        then upload the file to gcs bucket.
+        """
+        filename = self._format_filename(table, date)
+        self._save_to_csv(filename, results)
+
+        blob_name = self._format_blob_name(table, date)
+        self._save_to_gcs(filename, blob_name)
+
+
+def get_querylog_query(
+    databases: list[str], tables: list[str], start: datetime, end: datetime
```

Comment:
Are we going to pass multiple databases and tables to this API? The typing suggests so. But the query below suggests its going to be only 1. If there are going to be multiple databases and tables then maybe change the query to use IN clause.

Reply:
@nikhars This is maybe a confusing by product of the [clickhouse querylog](https://clickhouse.com/docs/en/operations/system-tables/query_log) - you have to query as an array of dbs or array of tables, but if you as for `WHERE tables in ['default.errors_dist', 'default.transactions_dist']` you will get queries that have both of those tables, not either. Same with databases.

I think it makes more sense for us to just query one table at a time, even if that means that some queries will run twice because they had multiple tables